af/shovel/centrifugation.py

```diff
@@ -771,8 +771,16 @@ def calc_measurement_flags(pgconn, flags_tbl, msm_tbl):
         ''')
         c.execute('''
         INSERT INTO {flags} SELECT msm_no, bool_or(anomaly), bool_or(confirmed) FROM (
-            SELECT msm_no, true AS anomaly, true AS confirmed FROM http_request_fp
-            WHERE msm_no IN (SELECT msm_no FROM {msm}) AND msm_no >= %s AND msm_no <= %s
+            SELECT http_request_fp.msm_no AS msm_no,
+                true AS anomaly,
+                true AS confirmed
+            FROM http_request_fp
+            JOIN fingerprint USING (fingerprint_no)
+            JOIN measurement_meta USING (msm_no)
+            JOIN report_blob USING (report_no)
```

Comment:
I'm unsure if it plays nicely with both invocations of `calc_measurement_flags()`. It probably does work for _ingestion_ case, but I'm unsure about reprocessing/reingestion case.

Reply:
Can you describe a bit more what happens in the case of reprocessing/reingestion?

Would any of these tables not exist in those cases?

Reply:
They will be probably empty WRT msm_no from `{msm}` table. It's unclear if it'll lead to flag being cleared or just ignored. I'd suggest to test that use-case as well if you consider that use-case useful.

Reply:
I see. I guess this problem would have presented itself also without my addition though, right?

Reply:
I don't think so.

Reply:
Can you give me some tips on how I might go about testing if the it poses some problems with reprocessing? I am not too familiar with how to operate the pipeline for reprocessing/reingestion.

Reply:
Sure. Sorry, it took me a while to understand what sort of tip may be useful :) Let me state some facts first:
- the query uses tables `measurement_meta` and `report_blob`
- `report_blob` is filled by `ReportFeeder`
- instance of `ReportFeeder` is created at `prepare_destination()`
- `ReportFeeder.report_row()` is called at `copy_data_from_autoclaved()` only `if code_ver is None` where `code_ver` comes from `iter_autoclaved_datum()`
- `core_ver` may be `NOT NULL` for measurements coming from `autoclaved` / `report` / `measurement` where `code_ver` in the table is less than `CODE_VER` declared in the code
- `NOTNULL(code_ver)` also calls `MeasurementExceptionFeeder` instead of `MeasurementFeeder` in `feeder = msm if code_ver is None else msm_exc`
- `calc_measurement_flags` called from `MeasurementExceptionFeeder` will get set of `msm_no` on "old" measurements (and only them)

It's unclear to me if joining "old" `msm_no` with "new" table `report_blob` will produce any output. I expect zero tuples to come out of that JOIN as I expect sets of `msm_no` values in these table to be non-intersecting.
So zero tuples will get `true AS confirmed` in `{flags}` table.
So I _expect_ `MeasurementExceptionFeeder.close()` to `UPDATE measurement` and drop all `confirmed` flags back to `NULL`. Maybe I'm wrong about that.

So, I suggest to test _at least_ the case when new table is added and reprocessing code path is triggered. The way I use to trigger that code path is ingesting a bucket locally, doing something like `UPDATE autoclaved SET code_ver = 1`, running `centrifugation.py` again. I'd suggest to compare flags set after the first and the second run. They _should_ be equal (as the data is not changed), but my understanding of the code makes me think that they will not be.

Reply:
Thanks for explaining this it's very useful!

I think there are some good bits of insight and knowledge that deserve going into the pipeline docs.

The last bit on updating the code_ver and re-running the centrifugation is especially useful and I am going to see if I can codify it as an integration test.

Reply:
`tsvdump` / `tsvdiffstat` were some of the proto-test scripts those may be useful for the case.

Reply:
According to the test the confirmed and unconfirmed count stays constant before and after reprocessing:
https://travis-ci.org/ooni/pipeline/builds/524104818#L787
https://travis-ci.org/ooni/pipeline/builds/524104818#L894

I still need to fully understand the implications of the text you are written above and wrap my head around it, but it seems like it passes the basic test.

Can you think of something more tricky that I can do to check more in depth if things are working as expected?

How can I test, for example, for re-ingestion?

Reply:
> count stays constant
Indeed. I'm a bit surprised and still don't understand why it works though. Maybe that's a symptom of some other unexpected behavior in the code...

I'd also suggest to consider checking _anomaly_ counts as well. The check comes at a cost of few lines of code and almost no extra runtime and these two flags have slightly different semantics and, maybe, something will be eventually caught.

> something more tricky
IDK. You can disable `FLAG_TRUE_TEMP` and inspect tables manually.

> re-ingestion
The easiest way to trigger re-ingestion code path is to `UPDATE autoclaved SET file_sha1 = digest('wat', 'sha1')`.

Reply:
I added a check for the anomaly flag value as well and it succeeds.

I also added a check that triggers re-ingestions and does the check there too and it also leads to consistent results.